# Helper scripts used during the hackathon `#AMRhack2021`

## Dataset download
* `get-url-list-from-ena.sh`: download a list of URLs from ENA, providing the study accession number (eg. PRJEB47011)
* `get-reads-from-tsv.sh`: download a list of FASTQs from ENA, providing the TSV file generated by `get-url-list-from-ena.sh`

## Dataset generation

### abricate-get-reads.py

Extract FASTQ reads from a BAM file (reads against contigs) and a TAB file produced by Abricate (ran on contigs).

```
usage: abricate-get-reads.py [-h] [-i DIR] [-a AMR] [-b BAM] [-g GENOME] [-o OUT_DIR] [--format {gene,accession,product}]
                             [--min-len MINLEN] [--min-cov MIN_COV] [--min-id MIN_ID] [--gene GENE] [-v]

Extract reads mapping to a ABRICATE AMR predictions.

Input: 
  * AMR predictions in tabular format
  * BAM file with reads mapped against the assembly

Output:
  * FASTQ file with reads mapping to the AMR predictions

optional arguments:
  -h, --help            show this help message and exit

Input files:
  -i DIR, --dir DIR     Directory with assembly, bam file and Abricate file
  -a AMR, --amr AMR     ABRICATE predictions in tabular format
  -b BAM, --bam BAM     BAM file with reads mapped against the assembly
  -g GENOME, --genome GENOME
                        FASTA file with the assembly [optional]

Output:
  -o OUT_DIR, --out-dir OUT_DIR
                        Output directory (will split FASTQ files by gene)
  --format {gene,accession,product}
                        Output filename [gene, product, accession]. Default: accession
  --min-len MINLEN      Skip reads shorter than MINLEN [default: 500]

Options:
  --min-cov MIN_COV     Minimum coverage in ABRICATE [90.0]
  --min-id MIN_ID       Minimum identity quality in ABRICATE [94.0]
  --gene GENE           Gene name to extract [None]
  -v, --verbose         Verbose mode
  ```